src/net/socket.rs

```diff
@@ -0,0 +1,229 @@
+use crate::{
+    config::Config,
+    error::{NetworkError, NetworkErrorKind, NetworkResult},
+    net::{connection::ActiveConnections, events::SocketEvent, link_conditioner::LinkConditioner},
+    packet::Packet,
+};
+use crossbeam_channel::{self, unbounded, Receiver, Sender};
+use log::error;
+use mio::{Evented, Events, Poll, PollOpt, Ready, Token};
+use std::{
+    self, io,
+    net::{SocketAddr, ToSocketAddrs},
+};
+
+const SOCKET: Token = Token(0);
+
+/// A reliable UDP socket implementation with configurable reliability and ordering guarantees.
+pub struct Socket {
+    socket: mio::net::UdpSocket,
+    config: Config,
+    connections: ActiveConnections,
+    recv_buffer: Vec<u8>,
+    link_conditioner: Option<LinkConditioner>,
+    event_sender: Sender<SocketEvent>,
+    packet_receiver: Receiver<Packet>,
+}
+
+impl Socket {
+    /// Binds to the socket and then sets up `ActiveConnections` to manage the "connections".
+    /// Because UDP connections are not persistent, we can only infer the status of the remote
+    /// endpoint by looking to see if they are still sending packets or not
+    pub fn bind<A: ToSocketAddrs>(
+        addresses: A,
+        config: Config,
+    ) -> NetworkResult<(Self, Sender<Packet>, Receiver<SocketEvent>)> {
```

Comment:
For future reference we should probably move these to some sort of `Socket` and `SocketHandle` where `SocketHandle` is clonable and provides the send and recv interface.  This would then allow you to use `Socket` to drive the inner transport and the `SocketHandle` to read and write. That should clean this up. `Socket` should still provide the read and write interface but is not clonable.

This is what Im thinking

```rust
struct Socket {
	fn send(&mut self, packet: Packet) -> Result<()>;
	fn recv(&mut self) -> Packet;
	fn start_polling(&mut self) -> Result<()>;
}

#[derive(Clone)]
struct SocketHandle {
	fn send(&mut self, packet: Packet) -> Result<()>;
	fn recv(&mut self) -> Packet;
}
```

mostly this is just me thinking out loud would love to hear more thoughts on this

Reply:
Actually I was thinking about this. Can do this now or in a future PR.

Reply:
let's do future, this was mostly me just dumping the thoughts out of my brain.

Reply:
I opened an issue for this #142